homeassistant/components/calendar/__init__.py

```diff
@@ -180,6 +191,171 @@ def _get_date(date):
+        self._cal_data['description'] = self.data.event.get('description', '')
+        self._cal_data['url'] = self.data.event.get('url', '')
         self._cal_data['start'] = start
         self._cal_data['end'] = end
         self._cal_data['all_day'] = 'date' in self.data.event['start']
+
+    def cleanup(self):
+        """Cleanup any start/end listeners that were setup."""
+        self._cal_data = {
+            'all_day': False,
+            'offset_time': 0,
+            'message': '',
+            'start': None,
+            'end': None,
+            'location': None,
+            'description': None,
+            'url': None,
+        }
+
+
+class EventData:
+    """Class to hold calendar data."""
+
+    def __init__(self, hass, persistence):
+        """Initialize the calendar."""
+        self.hass = hass
+        self.items = []
+        self.persistence = persistence
+
+    def _get_item(self, title, start, uid=None):
+        # Check if the object exists
+        for item in self.items:
+            if item.get('uid') is not None and uid == item['uid']:
+                return item
+            elif title == item['title'] and \
+                    self._get_date(start).isoformat() == item['start']:
+                # Item already created
+                return item
+        return None
+
+    @staticmethod
+    def _get_date(date):
+        """Get the dateTime from date or dateTime as a local."""
+        if 'date' in date:
+            return dt.start_of_local_day(dt.dt.datetime.combine(
+                dt.parse_date(date['date']), dt.dt.time.min))
+        return dt.as_local(dt.parse_datetime(date['dateTime']))
+
+    @callback
+    def async_add(self, title, start, end=None, url=None, color=None,
+                  all_day=False, description=None, location=None, uid=None):
+        """Add a calendar item."""
+        # Check if the object exists
+        if self._get_item(title, start, uid) is not None:
+            return
+        # Create new item
+        item = {
+            'title': title,
+            'start': start,
+            'all_day': all_day,
+            'id': uuid.uuid4().hex
+        }
+        for key in ('end', 'url', 'color', 'description', 'location', 'uid'):
+            value = locals().get(key)
+            if value is not None:
+                item[key] = value
+
+        item['start'] = self._get_date(item['start']).isoformat()
+        item['end'] = self._get_date(item['end']).isoformat()
+
+        self.items.append(item)
+        self.hass.async_add_job(self.save)
+        return item
+
+    @asyncio.coroutine
+    def async_load(self):
+        """Load items."""
+        def load():
+            """Load the items synchronously."""
+            return load_json(self.hass.config.path(self.persistence),
+                             default=[])
+
+        self.items = yield from self.hass.async_add_job(load)
+
+    def save(self):
```

Comment:
We should not save the items. We should query the platforms directly.

Reply:
ho ! :) Do you have an example how to do that ?!